src/session/session_op_tracker.c

```diff
@@ -0,0 +1,270 @@
+		WT_RET(__wt_session_op_tracker_clear(session));
+
+	if ((entry = TAILQ_FIRST(&session->op_tracker_availq)) != NULL) {
+		TAILQ_REMOVE(&session->op_tracker_availq, entry, aq);
+	} else
+		WT_RET(__wt_calloc_one(session, &entry));
+
+	/* We have a new entry object now, start setting it up. */
+	WT_RET(__wt_epoch(session, &entry->start));
+	memcpy(&entry->last_start, &entry->start, sizeof(struct timespec));
+	entry->api_boundary = api_boundary;
+	entry->depth = 1;
+	TAILQ_INSERT_TAIL(&session->op_trackerq, entry, q);
+
+	/*
+	 * Find the parent for this entry. It is the last entry in the queue
+	 * that hasn't been finished.
+	 */
+	for (prev_entry = TAILQ_PREV(entry, __op_tracker, q);
+	    prev_entry != NULL;
+	    prev_entry = TAILQ_PREV(prev_entry, __op_tracker, q)) {
+		if (prev_entry->done)
+			continue;
+		WT_ASSERT(session, type != 1009 || prev_entry->type != 11);
+		entry->start_offset_us =
+		    WT_TIMEDIFF(entry->start, prev_entry->start);
+		entry->depth = prev_entry->depth + 1;
+		prev_entry->self_time_us +=
+		    WT_TIMEDIFF(entry->start, prev_entry->last_start);
+		break;
+	}
+
+	/* Create and insert the first entry into the slow op tracker. */
+	entry->type = type;
+	*entryp = entry;
+
+	if (api_boundary)
+		++session->api_call_depth;
+
+	WT_ASSERT(session,
+	    !TAILQ_EMPTY(&session->op_trackerq));
+
+	return (0);
+}
+/*
+ * __wt_session_op_tracker_finish_entry --
+ *	Finalize an entry.
+ */
+int
+__wt_session_op_tracker_finish_entry(
+    WT_SESSION_IMPL *session, WT_OP_TRACKER_ENTRY *entry)
+{
+	WT_OP_TRACKER_ENTRY *prev_entry;
+
+	/*
+	 * Don't track operations completed by internal sessions. It's also
+	 * OK to ignore calls with a NULL entry - they are operations that
+	 * weren't fully created, and we handle finish with a NULL entry to
+	 * avoid complicating callers code.
+	 */
+	if (F_ISSET(session, WT_SESSION_INTERNAL) || entry == NULL)
+		return (0);
+
+	/*
+	 * Special case for connection close - skip handles that are no longer
+	 * valid. It would be nice to capture this elsewhere, but automatically
+	 * capturing all API calls makes that non-trivial.
+	 */
+	if (session->iface.connection == NULL)
+		return (0);
+
+	if (entry->api_boundary)
+		--session->api_call_depth;
+
+	WT_RET(__wt_epoch(session, &entry->end));
+	entry->done = 1;
+	/*
+	 * Set a timer in the parent if this operation is nested to facilitate
+	 * tracking self time.
+	 */
+	for (prev_entry = TAILQ_PREV(entry, __op_tracker, q);
+	    prev_entry != NULL;
+	    prev_entry = TAILQ_PREV(prev_entry, __op_tracker, q)) {
+		if (prev_entry->done)
+			continue;
+		memcpy(&prev_entry->last_start,
+		    &entry->end, sizeof(struct timespec));
+		break;
+	}
+
+	/* Reporting is done as we are returning from the API call. */
+	if (entry->api_boundary && session->api_call_depth == 0) {
+		/* Update the self timer since this is the end of the trace. */
+		entry->self_time_us +=
+		    WT_TIMEDIFF(entry->end, entry->last_start);
```

Comment:
WT_TIMEDIFF returns the time difference in nanoseconds so divide by 1000.  I've generally been dividing by WT_MILLION.  Do we need less than 1 msec granularity?